Add the age-structured model

I-Bouros opened the pull request:
Closes #4 partially.

Create the class based age structured SEIR model. Methods + tests.

Commit:
Add the age-structured model

Commit:
Upload libraries in the test file

Commit:
Adapt the imports

Commit:
Refactorise

Commit:
Add error for compartments not summing up to 1

oscardm20994 committed:
added snapshot testing

solveigvandervegt commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+         prototype = list(
+           name = NA_character_,
+           output_names = list('S', 'E', 'I', 'R', 'Incidence'),
+           parameter_names = list('S0', 'E0', 'I0', 'R0', 'a', 'b', 'c'),
```
currently we are using different names for the parameters. You guys are using a, b, and c and we (in the basic SEIR model) are using b(eta), k(appa), g(amma). It might be good for us to reach a consensus across the project on this. Happy to copy your parameter names if everyone is ok with that.

ben18785 replied:
Yep, think we should keep the names consistent if possible. I'd be tempted to go with the predominant literature parameter values: so "beta" for the transmission parameter and, if relevant, "mu" for the death parameter. Might be worth having a look at a few papers to see what they use but not crucial.

oscardm20994 replied:
Yep, I agree with going with the literature. Have changed ours to b, k and g.

solveigvandervegt commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+#' @param I0 initial fraction of the population that is infected
+#'           by age group.
+#' @param R0 initial fraction of the population that has recovered
+#'           by age group.
```
Maybe it would be good to note what data type your function can handle here? Since you are dealing with age groups, you'll probably have vectors of some sort as input and I think it might be useful to a user to know which data types the code can handle?

ben18785 replied:
Yep, agreed.

oscardm20994 replied:
added text signally that data can be added as vector of fractions (1 element per age group).

solveigvandervegt commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+        stop(glue('{p} storage format must be a vector.'))}
+    }
+    if(sum(S0, E0, I0, R0) != 1){
+      stop('All comparmtnets need to sum up to 1.')
```
typo -> compartments

oscardm20994 replied:
Have changed this!

solveigvandervegt commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+        #printtype of the plotting object to the command line for testing
+        print(typeof(SEIRplot))
+      }
+    }
```
I know this was briefly discussed in the meeting so maybe I am beating a dead horse but I am not quite convinced by the need for a plotting option within the simulate function. I get that we want a quick option to verify results but if we eventually want to write a separate function for "nice" plots that handles the output from any function, I think it would be just as easy to add a flag there that let's you switch between a basic and a more detailed plot. Plotting of any kind would then be consistent between models and we don't need to have this chunk of plotting code in every single class separately. I will copy this into our basic SEIR class in case we want to stick with it but I think it's something to consider :)

ben18785 replied:
Having thought about this, I think I agree with @solveigvandervegt. I think it'll be easier to make generic plotting methods that'll work between classes of SEIR model if we keep them separate from the classes themselves. Otherwise, you're likely to have, what is termed as a "inappropriate intimacy" code smell: https://refactoring.guru/smells/inappropriate-intimacy

oscardm20994 replied:
Thanks for raising this both. I see how it might be more trouble than it's worth to having plotting routines embedded into the definition of the class itself given that the user will most likely want to have more control over any plotting they do. I've removed the routine for now (i'll keep it somewhere locally or on slack in case we need it later).

solveigvandervegt reviewed (commented):
I think it looks really good :) just some small things to consider from me

solveigvandervegt commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+    output$Incidence <- unname(as.matrix(n_inc))
+
+    # Return only at integer times
+    return(output)
```
Is the incidence number something we want to include in the output?

ben18785 replied:
I'd say we probably do want to include as it's a really common output...

ben18785 commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+#' Defines an age-structured simple SEIR model.
+#' Slots of this class are as follows
```
I don't think you need this line "Slots of this class.."

oscardm20994 replied:
yep, removed

oscardm20994 replied:
yep good point, removed!

ben18785 commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+#'     which are used by the model.
+#' @slot parameters named list containing parameters of the model.
+#'     Initial values for each compartment, S0, E0, I0, R0.
+#'     Other parameters: a, b, c which represent the rates of changes
```
See @solveigvandervegt's comment below about these parameter names. Also you probably want to explain what these parameters actually mean. Is there a way we could write the equations of the system in the docstrings somewhere?

oscardm20994 replied:
I've added the set of equations in the docstring assocaited with the "simulate" function. They can be moved up to the docstrings associated with the class definition if we want. We don't have the implementation of contact matrices so currently the equations are just the SEIR system for a single age group with not cross age interactions. We will work to implement this ASAP.

ben18785 replied:
Cool -- thanks. Yes, I think it'd be good to add age-structured contact matrices. How about just using the POLYMOD matrices for starters?

ben18785 commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+#' @import glue
+#' @import ggplot2
+
+setClass('age_model',
```
Do we want a more descriptive name like "SEIRAge"?

oscardm20994 replied:
yep good point, switched class name to SEIRAge

oscardm20994 replied:
yep, class name changed

ben18785 replied:
Thanks!

ben18785 commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+         )
+)
+
+#' @describeIn Retrieves parameters for an age-structured simple SEIR model.
```
I'm not sure this is quite how `describeIn` is used. I may be wrong but I think you need to say where the documentation is going to be housed for this function: https://r-pkgs.org/man.html#multiple-man

ben18785 commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+#' @describeIn Retrieves parameters for an age-structured simple SEIR model.
+#'
+#' @param object An object of the class age_model.
+setGeneric('get_parameters', function(object) standardGeneric('get_parameters'))
```
I'd be tempted to change it from `get_parameters` to `transmission_parameters`. You might also make a setter and a getter for it as described here: https://adv-r.hadley.nz/s4.html

oscardm20994 replied:
we use get_parameters as it  gets the initial population fractions as well as transmission parameters but more than happy to change if you think it's best. Alternatively could separate out get functions for initial fractions and transmission parameters (something like initials() & transimssion_parameters()).

ben18785 replied:
Up to you: I'd be tempted to separate them as there will be quite a few parameters, and it'd be good to avoid a long list. But pretty each about this point.

ben18785 commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+         prototype = list(
+           name = NA_character_,
+           output_names = list('S', 'E', 'I', 'R', 'Incidence'),
+           parameter_names = list('S0', 'E0', 'I0', 'R0', 'a', 'b', 'c'),
```
I'd be tempted to split out the initial conditions from the "transmission parameters", perhaps? I think this might make things neater (especially when we start adding more parameters to the model).

ben18785 commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+  function(object) object@parameters
+)
+
+#' @describeIn Sets parameters for an age-structured simple SEIR model.
```
As I mention above, I think making a combined setter and getter in a single method makes most sense, perhaps? (It makes the package namespace much smaller in any case.)

ben18785 commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+#' @return updated version of the age-structured SEIR model.
+setGeneric(
+  'set_parameters',
+  function(object, S0, E0, I0, R0, a = 1, b = 1, c = 1){
```
I'm not sure you want default values in a setter. I'd just make them all required.

oscardm20994 replied:
agreed, default values removed

ben18785 commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+      stop('Invalid initial conditions. Must add up to 1.')
+    }
+
+    #create list of parameter values
```
A minor thing, but I'd try to put spaces between the hash and the start of the comment as it makes readability easier.

oscardm20994 replied:
This has been implemented throughout

ben18785 replied:
Thanks!

ben18785 commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+        stop(glue('Wrong number of age groups for {p}
+              compartments.'))}
+      if(!is.double(params[[p]])){
+        stop(glue('{p} storage format must be a vector.'))}
```
I'm not sure I understand the error message here: if the parameters aren't doubles does that mean it's not a vector?

ben18785 replied:
Also, I'm not sure whether checking `is.numeric` might make more sense here as it's more general?

oscardm20994 replied:
yep good point, we don't test whether if it's a vector here, just the type of the elements. error message changed to "format must be numeric" and check changed to is.numeric

ben18785 replied:
Cool, thanks

ben18785 commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    #check format of parameters a and b
+    if(any(length(a) != 1 | length(b) != 1 | length(c) != 1)){
```
Do you need the `any` here?

oscardm20994 replied:
good point, I think it's superfluous

ben18785 replied:
Great, thanks!

ben18785 commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+
+    #check format of parameters a and b
+    if(any(length(a) != 1 | length(b) != 1 | length(c) != 1)){
+      stop('The rates of change between compartments are 1-dimensional.')
```
Shouldn't it be that the rates are _not_ 1-dimensional? I'd be tempted to rephrase the message as, "The parameter values should each be one-dimensional."

oscardm20994 replied:
yep, well spotted. Changed

ben18785 replied:
Thanks

ben18785 commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+           }
+)
+
+#' @describeIn Solves an age-structured simple SEIR model.
```
I'm not sure the "simple" is needed here.

ben18785 replied:
I think you may also want to include the method of solution used (i.e. numerical integration). I think it might also be good if we could allow a user to change the default numerical integration method used by deSolve (as think this might link quite well with some of the inference work that you're doing).

oscardm20994 replied:
Removed simple and described the ability of the class to solve the set of odes associated with the system through numerical integration of a chosen method. 

Also added argument to the simulate_SEIRAge function which is a string specifying the numerical solver. Default value is 'lsoda' which is the default for deSolve's ode function.

ben18785 replied:
Parfait, thanks.

ben18785 commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    # compute incidence number
+    total_inf <- output[, (2*age+2):(3*age+1)] + output[, (3*age+2):(4*age+1)]
```
I'm not sure I understand what this code is doing here. I'm also a little wary of these hard-coded numeric indices: is there a way to make them strings so as to a) make it more readable and b) avoid issues with these hard coded values?

ben18785 replied:
See my comment below about returning an output in a different form, which would probably make it easier to avoid this type of hard-coded values issue.

ben18785 commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+          dS <- -a*S*I
+          dE <- a*S*I - b*E
+          dI <- b*E - c*I
+          dR <- c*I
```
I think I'd be tempted to solve for cumulative incidence here via `dCdt = beta * S * I`. Could we also add deaths (from the infected compartment)?

ben18785 commented on tests/testthat/_snaps/age_model_class.md:
```diff
@@ -0,0 +1,16 @@
+
+    [1] "list"
+    [1] "list"
+      time  S1  S2 E1 E2 I1 I2 R1 R2 Incidence.1 Incidence.2
```
How about we return the output in long form. So we'd have one column that is time, one that is compartment (S, E, I, R, D, and some letter for incidence), one for age group which gives the actual age range being considered and another for value? That would save us having to use the more arbitrary S1, S2 etc...

ben18785 replied:
It also makes it easier for plotting.

ben18785 reviewed (changes requested):
Thanks @I-Bouros and @oscardm20994 -- a really good start. Forgive my somewhat involved commenting but I think this one's especially important to get right given the dependence of the other models on this one.

FarmHJ commented on R/age_model_class.R:
```diff
@@ -0,0 +1,222 @@
+#'           by age group.
+#' @param E0 initial fraction of the population that has been exposed
+#'           by age group.
+#' @param I0 initial fraction of the population that is infected
```
If I'm not wrong, in SEIR model, exposed individuals are infected, but not infectious. Would it be better to change compartment I to infectious, instead of infected?

oscardm20994 replied:
Good point, have made this edit

FarmHJ reviewed (commented):
Looks great, just minor comments from me. Most of it is covered by @solveigvandervegt and @ben18785. I realised a Rplots.pdf file, under the testthat folder, is uploaded. Is it for testing purposes? Do we want to upload the file or do we want to ignore it in say .gitignore?

oscardm20994 committed:
changed parameter names to b, k, g

oscardm20994 committed:
Update test_age_model_class.R

oscardm20994 committed:
Update age_model_class.R

oscardm20994 committed:
Update age_model_class.R

oscardm20994 committed:
removed plotting routine

oscardm20994 committed:
changed simulate function name, added solve_method argument

Commit:
Change names of parameters in generic of set_parameters

oscardm20994 committed:
Merge branch 'i4-age-structure' of https://github.com/Como-DTC-Collaboration/como-models into i4-age-structure

Commit:
Eliminate default values

Commit:
Space & Hashes for comments

oscardm20994 committed:
Update age_model_class.R

oscardm20994 committed:
Merge branch 'i4-age-structure' of https://github.com/Como-DTC-Collaboration/como-models into i4-age-structure

oscardm20994 committed:
remove "any" from checks on parameter lengths

Commit:
Split parameters and methods

Commit:
Update docstrings

Commit:
Fixed docstring?

FarmHJ commented on R/SEIRAge.R:
```diff
@@ -0,0 +1,275 @@
+      if(!is.numeric(ic[[p]])){
+        stop(glue('{p} format must be numeric'))}
+    }
+    if(sum(S0, E0, I0, R0) != 1){
```
The sum of initial conditions equals one is checked twice.

FarmHJ commented on R/SEIRAge.R:
```diff
@@ -0,0 +1,275 @@
+#' All rates of change between compartments are equal regardless of
+#' age group.
+#'
+#' @return Updated version of the age-structured SEIR model.
```
Maybe provide more information on what the function returns.

ben18785 replied:
I agree: it's also a bit unclear what "Updated" means here?

chonlei replied:
Agree too. See comment below, maybe add a reference as to what is updated from what.

FarmHJ commented on R/SEIRAge.R:
```diff
@@ -0,0 +1,275 @@
+#'
+#' All initial conditions must sum up to 1.
+#'
+#' @return Updated version of the age-structured SEIR model.
```
Maybe provide more information on what the function returns.

FarmHJ commented on tests/testthat/test_age_model_class.R:
```diff
@@ -0,0 +1,127 @@
+  # Test input errors
+  expect_error({
+    simulate_SEIRAge(my_model, '0')
+    simulate_SEIRAge(my_model, seq(0, 2, by = 1))})
```
What are you trying to test here? This test seems correct to me. It looks like it is giving an 'or' output of all tested statements.

FarmHJ commented on R/SEIRAge.R:
```diff
@@ -0,0 +1,275 @@
+#' ordinary differential equations of the model with a chosen method of
+#' numerical integration.
+#'
+#' @slot name character representing name of model
```
I think @ben18785 mentioned this before, you might want to remove the 'name' slot, it is not used in the class.

FarmHJ commented on R/SEIRAge.R:
```diff
@@ -0,0 +1,275 @@
+#' @param R0 initial fraction of the population that has recovered
+#'           by age group. Data can be provided as a list or vector of doubles.
+#'
+#' All initial conditions must sum up to 1.
```
This is repeated in the description before @param.

FarmHJ reviewed (commented):
Sorry for the extra comments. I had overlooked previously. I saw it while trying to follow the code structure for the SEIR model.
A minor suggestion, you might want to delete the tests/testthat/Rplots.pdf file. I believe it was uploaded when you had the plotting function in your model.

chonlei commented on R/SEIRAge.R:
```diff
@@ -0,0 +1,275 @@
+#' \deqn{\frac{dS_i(t)}{dt} = - b S_i(t) I_i(t)}
+#' \deqn{\frac{dE_i(t)}{dt} =  b S_i(t) I_i(t) - k E_i(t)}
+#' \deqn{\frac{dI_i(t)}{dt} = k E_i(t) - g I_i(t)}
+#' \deqn{\frac{dR_i(t)}{dt} = g I_i(t)}
```
Not sure if it's a common thing to do in R, but maybe we can give some reference for the model?

ben18785 replied:
Yep, good idea.

fcooper8472 committed:
Try with error on error not warning

oscardm20994 committed:
added age_range slot and wide format output

oscardm20994 committed:
Merge branch 'i4-age-structure' of https://github.com/Como-DTC-Collaboration/como-models into i4-age-structure

oscardm20994 committed:
altered tests for new output format

oscardm20994 committed:
changes to testing

oscardm20994 committed:
intorduced contact matrices

oscardm20994 committed:
more contact implementations

oscardm20994 committed:
Update SEIRAge.R

ben18785 committed:
Made class exported

ben18785 committed:
Spaced equations

ben18785 committed:
Added contact matrices, cleaned up tests and setters / getters

ben18785 committed:
Added various imports to SEIR Age functions

ben18785 committed:
First very rough draft of vignette

oscardm20994 committed:
added realistic demography for vignette simulations

oscardm20994 committed:
added initial condition tests

oscardm20994 committed:
Update SEIRD_age_structured.Rmd

oscardm20994 committed:
added death compartment

Commit:
Rerun documetation

Commit:
Merge branch 'main' into i4-age-structure

Commit:
Add death to vihgnette

Commit:
Solve issue with generic

Commit:
Update name of class in vignette

Commit:
Rename classes and files

Commit:
Add load files

fcooper8472 commented on R/SEIRDAge.R:
```diff
@@ -0,0 +1,328 @@
+    R0 = value$R0
+    D0 = value$D0
+    # check that ICs are valid
+    if (sum(S0, E0, I0, R0, D0) != 1) {
```
This looks like a floating point comparison? If so should this just be within some defined tolerance?

fcooper8472 commented on R/SEIRDAge.R:
```diff
@@ -0,0 +1,328 @@
+      if(!is.numeric(ic[[p]])){
+        stop(glue('{p} format must be numeric'))}
+    }
+    if(sum(S0, E0, I0, R0, D0) != 1){
```
Ditto on floating point comparison

fcooper8472 commented on R/SEIRDAge.R:
```diff
@@ -0,0 +1,328 @@
+  function(object, times, solve_method = 'lsoda') {
+
+    # error if times is not a vector or list of doubles
+    if(!is.double(times)){
```
Given that the default is `seq(0, 100, by = 1)` it might be reasonable to check that times is just numeric rather than specifically double?  And perhaps even check that it's monotonically increasing?

fcooper8472 commented on R/SEIRDAge.R:
```diff
@@ -0,0 +1,328 @@
+
+    # melt the incidence dataframe to long format
+    incidence_temp = melt(n_inc, id.vars=NULL)
+    print(incidence_temp)
```
Is this print intended?

I-Bouros replied:
I tried to fins sth for this one, but could not find anything useful for the docstring part. If it were code, a ``paste0`` would have worked

fcooper8472 commented on R/data.R:
```diff
@@ -0,0 +1,60 @@
+#' @format A Large list of 152 tibbles, each a 16x16 list of doubles.
+#'
+#' @source \url{https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005697#sec020}
+"contact_other"
```
Generally advisable to have a newline at the end of every file.

fcooper8472 commented on R/data.R:
```diff
@@ -0,0 +1,60 @@
+#'
+#' @format A Large list of 152 tibbles, each a 16x16 list of doubles.
+#'
+#' @source \url{https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005697#sec020}
```
Is it possible to split these URLs over 2 lines?

fcooper8472 commented on vignettes/SEIRD_age_structured.Rmd:
````diff
@@ -0,0 +1,236 @@
+```
+
+
+Now we simulate both models and plot the Infections over time. We can see that for the model initialised with infections among 15-30 years olds, the peak of infections appears earlier and dies away quicker than the model initialised with infections in 50-70 year olds. Despite this difference, the magnitude of the infection peaks in both bases are comparable as well as the distribution of infectio ...
````
lowercase i on Infections

fcooper8472 reviewed (commented):
I have left a few very minor comments, but over all it's looking very good.

fcooper8472 commented:
There are also a couple of checks that aren't covered by a test yet:
https://app.codecov.io/gh/Como-DTC-Collaboration/como-models/compare/5/tree/R/SEIRDAge.R

Commit:
Solve issues in PR

Commit:
Update tests

rccreswell commented on R/SEIRDAge.R:
```diff
@@ -0,0 +1,333 @@
+          # rate of change
+          dS <- -b * S * C %*% I
+          dE <- b * S * C %*% I - k * E
+          dI <- k * E - g * I
```
```suggestion
          dI <- k * E - g * I - mu * I
```
Do we need to account for I -> D? (also in the documentation above)

I-Bouros replied:
You are right! the RHS of sum of the equations of the system must be 0 to keep the total population constant in time

Commit:
Correct error in model assumption

ben18785 committed:
Small changes to vignette text

ben18785 committed:
Removed space

ben18785 committed:
Changed SEIR to SEIRD

Commit:
Patially deal with the documentation

oscardm20994 committed:
changed initial conditions in vignette

oscardm20994 committed:
Merge branch 'i4-age-structure' of https://github.com/Como-DTC-Collaboration/como-models into i4-age-structure

Commit:
Separate output into states and changes

Commit:
Update the vignette as well

oscardm20994 committed:
documentation for demographic data

oscardm20994 committed:
documentation of pop data

oscardm20994 committed:
documentation of pop data

Commit:
Correct error in vignette

Commit:
Solve documentation issue

ben18785 committed:
changed incidence factoring to match state factoring

also removed unneeded packages

ben18785 committed:
Corrected how incidence and deaths are calculated

- added cumulative incidence state to handled incidence
- also changed tests so that they accommodate these changes
- changed death -> deaths in the output to match SEIRD
- added back in dplyr, forcats, ggplot2 as dependencies

ben18785 committed:
small changes to vignette

ben18785 committed:
Corrected issue with vignette
